crates/agent/src/agent_profile.rs

```diff
@@ -0,0 +1,283 @@
+use std::sync::Arc;
+
+use agent_settings::{AgentProfileId, AgentProfileSettings, AgentSettings};
+use assistant_tool::{Tool, ToolSource, ToolWorkingSet};
+use gpui::{App, Entity};
+use settings::Settings;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AgentProfile {
+    id: AgentProfileId,
+    tool_set: Entity<ToolWorkingSet>,
+}
+
+impl AgentProfile {
+    pub fn new(id: AgentProfileId, tool_set: Entity<ToolWorkingSet>) -> Self {
+        Self { id, tool_set }
+    }
+
+    pub fn id(&self) -> &AgentProfileId {
+        &self.id
+    }
+
+    pub fn enabled_tools(&self, cx: &App) -> Vec<Arc<dyn Tool>> {
+        let Some(settings) = self.settings(cx) else {
+            return vec![];
+        };
+        self.tool_set
+            .read(cx)
+            .tools(cx)
+            .into_iter()
+            .filter(|tool| Self::is_enabled(&settings, tool.source(), tool.name()))
+            .collect()
+    }
+
+    fn is_enabled(settings: &AgentProfileSettings, source: ToolSource, name: String) -> bool {
+        match source {
+            ToolSource::Native => *settings.tools.get(&Arc::from(name)).unwrap_or(&false),
+            ToolSource::ContextServer { id } => {
+                if settings.enable_all_context_servers {
+                    return true;
+                }
+
+                let Some(preset) = settings.context_servers.get(&Arc::from(id)) else {
+                    return false;
+                };
+                *preset.tools.get(&Arc::from(name)).unwrap_or(&false)
+            }
+        }
+    }
+
+    fn settings(&self, cx: &App) -> Option<AgentProfileSettings> {
+        AgentSettings::get_global(cx)
```

Comment:
I don't cache much on the profile to avoid having to manage as much state, since we are wrapping the ToolWorkingState anyway

However, we could precalculate some of this stuff and only change it on settings changes if that feels better?